homeassistant/components/ecobee/switch.py

```diff
@@ -0,0 +1,105 @@
+from typing import Any
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import EcobeeData
+from .const import DOMAIN
+from .entity import EcobeeBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DATE_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+@dataclass(frozen=True, kw_only=True)
+class EcobeeSwitchEntityDescription(SwitchEntity):
+    """Class describing Ecobee switch entities."""
+
+    ecobee_setting_key: str
+    set_fn: Callable[[EcobeeData, int, int], Awaitable]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ecobee thermostat switch entity."""
+    data: EcobeeData = hass.data[DOMAIN]
+    _LOGGER.debug("Adding ventilators 20 min switch if present")
+
+    async_add_entities(
+        (
+            EcobeeVentilator20MinSwitch(data, index)
+            for index, thermostat in enumerate(data.ecobee.thermostats)
+            if thermostat["settings"]["ventilatorType"] != "none"
+        ),
+        True,
+    )
+
+
+class EcobeeVentilator20MinSwitch(EcobeeBaseEntity, SwitchEntity):
+    """A Switch class, representing 20 min timer for an ecobee thermostat with ventilator attached."""
+
+    def __init__(
+        self,
+        data: EcobeeData,
+        thermostat_index: int,
+    ) -> None:
+        """Initialize ecobee ventilator platform."""
+        super().__init__(data, thermostat_index)
+        self._attr_unique_id = "ventilator_20m_timer"
+        self._attr_native_value = False
+        self._time_zone_delay = datetime.strptime(
+            self.thermostat["utcTime"], DATE_FORMAT
+        ) - datetime.strptime(self.thermostat["thermostatTime"], DATE_FORMAT)
+
+    @property
+    def is_on(self) -> bool:
+        """Get the latest state from the thermostat."""
+        return self._attr_native_value
+
+    async def async_update(self) -> None:
+        """Get the latest state from the thermostat."""
+        await self.data.update()
+
+        ventilatorOffDateTime = self.thermostat["settings"]["ventilatorOffDateTime"]
+
+        self._attr_native_value = (
+            ventilatorOffDateTime is not None
+            and ventilatorOffDateTime != ""
+            and datetime.strptime(ventilatorOffDateTime, DATE_FORMAT)
+            + self._time_zone_delay
+            >= datetime.now()
+        )
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Set ventilator 20 min timer on."""
+        await self.hass.async_add_executor_job(
+            self.data.ecobee.set_ventilator_timer, self.thermostat_index, True
+        )
+        self._attr_native_value = True
+        self.thermostat["settings"]["ventilatorOffDateTime"] = (
+            datetime.now() + timedelta(minutes=20) - self._time_zone_delay
```

Comment:
Why don't we create a time entity for this? Gives way more flexibility to the user

Reply:
so we don't really have a chance to set it. 

TLDR: this specific line is only changing local data, not sending to Ecobee Server. The only thing you can send is `set_ventilator_timer` which set a 20 min (fix) timer. 


Context:

The real setting is `set_ventilator_timer(id, True)` this will force the Ecobee server to set `ventilatorOffDateTime` at a value of now + 20 min. 

The reason we do this set `self.thermostat["settings"]["ventilatorOffDateTime"]` to an approximate value, is to get the local value in Home Assistant so the switch don't flick back on/off until an update is done.

Reply:
no, not currently implemented. 

Also. If this would be implemented there would be 2 things. 

1) if the 20 min timer switch is on or off
2) if the ventilator is on or off

The ventilator can be on for other reasons than the 20 min switch being on. (exemple the `min ventilator time` number in the integration) 

Basically the way the device works is. You can configure it to be on x min/hour. `min ventilator time`  and then the ecobee turns it on/off to match that number. 

But if you want to give a boost of ventilator, you can turn on the `20 min switch` on the ecobee interface to run it for the next 20 min (regardless of the `min ventilator time` ). The ecobee behaviour is that switch stays on for that 20 min and go off after. 

This last piece is what I am adding today.

Reply:
For what's it's worth, if we wanted a binary_sensor to say if the ventilator is on/off it would be done by another data point call `enabled equipement` which is a coma separated list of all the things on. ex: `fan,ventilator,heat` 

The reason I assume nobody ever modeled it, is because of the bad refresh rate of the ecobee device (3 min) it makes it very un-usable

Reply:
Yea but that's what I meant with the question if we can see the switch being on in other entities, and if I understand correctly, we can

Reply:
the other entity would be on `enabled equipement` and we could not distinguish if the 20 min timer is on. (vs just the ventilator)

Reply:
> Would it maybe fit to be a button instead to avoid keeping state ourselves?

Yes, that seems like the best solution.  As we'd want the state to be consistent regardless of where the action is taken.  Meaning if the 20 minute vent is initiated / canceled from the panel or HA the state should be the same.   If the ecobee can provide remaining ventilation duration then this could be done; but it appears it does not?  If so, the best is a button to initiate and a binary sensor to report ventilation State.

Reply:
So the value that we set here get configured by Ecobee servers regardless where the switch is flipped. 

In the device or in HA. 

The reason we modified that value was for getting a better responsiveness. I will remove the change to this.

Reply:
The thing with a button will be in-ability to turn off the 20 min timer. This is why I think the switch makes more sense.

Reply:
So if the 20 minute ventilate is set on the panel.  Then the value of 

```python
self.thermostat["settings"]["ventilatorOffDateTime"]
``` 

will contain the off time?

Reply:
Yes exactly

Reply:
No.

this line: 
`self.thermostat["settings"]["ventilatorOffDateTime"] = (datetime.now() + timedelta(minutes=20) - self._time_zone_delay`

is never leaving HA. it was only added to enhance user experience. 

The only thing that leaves HA is `set_ventilator_timer(id, True)` (or false)

When Ecobee receive that command they:
- Start the ventilator 
- configure thermostat["settings"]["ventilatorOffDateTime"] to now+20 min. 
- when that delay is done, they turn off the ventilator. 


we could have skipped the `self.thermostat["settings"]["ventilatorOffDateTime"] = (datetime.now() + timedelta(minutes=20) - self._time_zone_delay` this is not in any way setting the real value it's just an estimate so HA has a temporary value until we get a refreshed value (up to 3 min)

Reply:
Now I understand, that line is an anti-pattern. So I will remove it. But it will not change the behavior of this new switch. (simply the refresh time of it)

Reply:
and if we send `set_ventilator_timer(id, False)` then it stop then Ecobee server stop the ventilator and set `ventilatorOffDateTime` to ""

Reply:
Got it, apologies for it taking so long.  To summarize

a) there is an api command to turn the "20 minute ventilator" (20mv)there are no other options for different time amount.
b) there is an api command to stop an in progress 20mv
c) the remaining time on the 20mv is reported by the thermostat, the update period is 3 minutes
d) the purpose of setting the remaining time in the code above was to make the switch be on while the system is waiting for the update in (c) to come through. 

To make it consistent with the panel UI, we'd want a way to turn the 20mv on / off, display whether it is on or off and display the remaining time in minutes?

Regarding (d), I can see why this is useful in order to prevent this scenario:

1) user initiates 20mv
2) there is no UI feedback that it succeeded (or the UIs optimistic switch reverts back to off after a timeout)
3) user does 1 again and again and again...
4) eventually 3 minutes later HA shows the ventilator running 

A question, what happens for other control points in the integration?  For example setting temperature setpoints, hvac modes, etc.   Are these reported faster? Are they set optimistically by the integration while waiting for an update? What's the current convention?  Sometime when there are technical limitations (like poll times), the best we can do is have a consistent convention.

Reply:
a) exact
b) exact 
c) exact (3 min is a mix of HA polling rate + very bad API from Ecobee) 
d) exact 


other input for this integration)
- numbers: After you change the value, it flips to the old value for few min (until an update) then you get the value you wanted. 
- Climate: as I look I found their workaround. 

I will try this: 
https://github.com/home-assistant/core/blob/6bff0c384ffb1d26d21b52b19eb92439549beb28/homeassistant/components/ecobee/climate.py#L361C7-L364C49

Reply:
I just pushed a new commit that replicate the logic/behaviour of the climate platform and it works great (After this PR, I will make one to fix the number platform) 

What do you think

Reply:
> To make it consistent with the panel UI, we'd want a way to turn the 20mv on / off, display whether it is on or off and display the remaining time in minutes?

To make it consistent with the panel, I want the switch to be on for as long as the 20 min timer is on. This switch shows if the timer is on, not if the ventilator is on.

Reply:
It's a whole lot cleaner this way.  @joostlek - what do you think?

The only functional item I'd like to see, to get the same functionality as the panel is how many minutes / seconds are remaining.  Consider adding an attribute to the switch that contains the remaining seconds or minutes?  I think this will be easier to use from a display than the time stamp which requires some conversion? Before doing that let's see what @joostlek thinks about the current progress / direction.

Reply:
>how many minutes / seconds are remaining.

This is NOT available on the device. It only says `Ventilator (20 Min)` and the value does not change. It does not go 19-18-17-16 etc. It is just a static label.